### PR TITLE
:seedling: Set providerID in phase ensure-provisioned

### PR DESCRIPTION
**Pull request by janiskemper**
**What this PR does / why we need it**:
Currently, we set the providerID of a baremetal server only after we confirmed that cloud init has successfully completed. However, CAPI also reacts on the result of cloud init, so there might be a temporary state when CAPI already reconciles the node object, CSRs are created, etc., but we have not yet set the providerID.

This leads for example to errors in the csr-controller, where we rely on the providerID to be set.

This PR sets the providerID already in ensure-provisioned state, so that these temporary errors should not occur anymore.

I think we will prefer this approach: https://github.com/syself/cluster-api-provider-hetzner/pull/1357

**TODOs**:
- [x] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: pkg/services/baremetal/baremetal/baremetal.go [hunk 1]
```diff
@@ -673,17 +673,27 @@ func (s *Service) setProviderID(ctx context.Context) error {
 		return fmt.Errorf("host not found for machine %s: %w", s.scope.Machine.Name, err)
 	}
 
-	if host.Spec.Status.ProvisioningState != infrav1.StateProvisioned {
+	// set providerID in phase ensure provisioned and provisioned as from this point on the node might join the cluster
```
Comment (guettli) [inline pkg/services/baremetal/baremetal/baremetal.go:676]: @janiskemper I suggest to use quotes to make it more readable: 

> set providerID in phase "ensure provisioned" and "provisioned" ...

Comment (janiskemper) [reply pkg/services/baremetal/baremetal/baremetal.go]: fixed

**Comment by guettli** [comment]
Just for the records, we cherry-picked this PR, and then the bm machines failed to join the cluster.

**Comment by guettli** [comment]
This was merged instead: https://github.com/syself/cluster-api-provider-hetzner/pull/1357

**Pull request closed without merging**